server/main_server/src/rodsServer.cpp

```diff
@@ -958,81 +1000,94 @@ int main(int argc, char** argv)
-                    if (status >= 0) {
-                        break;
-                    }
+    const auto launch_agent_factory = [&] {
+        try {
+            int status;
+            int w = waitpid(agent_spawning_pid, &status, WNOHANG);
+            if (w != 0) {
+                ix::log::server::info("Starting agent factory");
+                auto new_pid = fork();
+                if (0 == new_pid) {
+                    close(pid_file_fd);
+
+                    ProcessType = AGENT_PT;
+
+                    // This appeared to be the best option at balancing cleanup and correct behavior,
+                    // however this may not perform the full cleanup that would happen on a normal return from main.
+                    // The other alternative considered here was throwing the code, however this was complicated
+                    // by the usage of catch(...) blocks elsewhere.
+                    exit(runIrodsAgentFactory(local_addr));
+                }
+                else {
+                    close(agent_conn_socket);
+                    ix::log::server::info("Restarting agent factory");
+                    agent_conn_socket = socket(AF_UNIX, SOCK_STREAM, 0);
+
+                    std::time_t sock_connect_start_time = std::time(nullptr);
+                    while (true) {
+                        const auto len = sizeof(local_addr);
+                        ssize_t status = connect(agent_conn_socket, (const struct sockaddr*) &local_addr, len);
+                        if (status >= 0) {
+                            break;
+                        }
+
+                        int saved_errno = errno;
+                        if ((std::time(nullptr) - sock_connect_start_time) > 5) {
+                            rodsLog(LOG_ERROR, "Error connecting to agent factory socket, errno = [%d]: %s",
+                                    saved_errno, strerror(saved_errno));
+                            exit(SYS_SOCK_CONNECT_ERR);
+                        }
 
-                    int saved_errno = errno;
-                    if ((std::time(nullptr) - sock_connect_start_time) > 5) {
-                        rodsLog(LOG_ERROR, "Error connecting to agent factory socket, errno = [%d]: %s",
-                                saved_errno, strerror(saved_errno));
-                        exit(SYS_SOCK_CONNECT_ERR);
                     }
 
+                    agent_spawning_pid = new_pid;
                 }
-
-                agent_spawning_pid = new_pid;
             }
         }
-    };
+        catch (...) {
+            // Do not allow exceptions to escape the CRON task!
+            // Failing to do so can cause the CRON manager thread to crash the server.
+        }
+    }; // launch_agent_factory
 
-    start_agent_server();
-    agent_watcher.interval(5).task(start_agent_server);
+    launch_agent_factory();
+    ix::cron::cron_builder agent_watcher;
+    agent_watcher.interval(5).task(launch_agent_factory);
```

Comment:
little surprised that the pattern is to launch, then watch(launch)... rather than just watch(launch).    might need a comment here if this is best/good for why.

Reply:
This is how it was before. It is only highlighted due to the lambda being renamed.

Reply:
sure.   just... would it work without the first launch?

Reply:
I believe it would be fine. When the agent factory is forked, the CRON task wraps the call with `exit()` (e.g. `exit(run_agent_factory())`). This is safe as long as no exception escapes `run_agent_factory()`.

And now that I've said it out loud, I'll modify that code to catch any exception that is thrown and guarantee that `exit()` is called with a suitable return code.

Reply:
Also, without manually calling `launch_agent_factory()`, there would likely be a delay in when the server can answer client requests.

Reply:
okay..... can we quantify that delay?     if it's within a second or two for the machinery to happen, that's probably fine.   if it's one full cron-wait() cycle, that could be too long...

Reply:
aka 'while' vs 'do while'

Reply:
Once I have things working, we can run some tests to see how quickly the server stands up.